src/wallet/api.rs

```diff
@@ -1060,48 +1074,177 @@ impl Wallet {
+
+        // Get next external and internal addresses using dedicated wallet APIs
+        let external_address = self.get_next_external_address()?;
+        let internal_address = self.get_next_internal_addresses(1)?[0].clone();
+
+        let external_address_script_pubkey = ScriptBuf::new_p2pkh(
+            &external_address
+                .pubkey_hash()
+                .ok_or_else(|| WalletError::General("Could not get pubkey hash".to_string()))?,
+        );
 
-        // Simple largest-first coin selection.
-        for unspent in unspents {
-            if remaining.checked_sub(unspent.0.amount).is_none() {
-                selected_utxo.push(unspent);
-                break;
-            } else {
-                remaining -= unspent.0.amount;
-                selected_utxo.push(unspent);
+        let internal_address_script_pubkey =
+            ScriptBuf::new_p2pkh(&internal_address.pubkey_hash().ok_or_else(|| {
+                WalletError::General("Could not get change pubkey hash".to_string())
+            })?);
+
+        // Prepare CoinSelectionOpt: Calculate required weights and costs, additional metrics
+        let long_term_feerate = 10.0;
+        let change_weight = Weight::from_vb_unwrap(
+            (Amount::SIZE // 8 bytes for amount
+                    + VarInt::from(internal_address_script_pubkey.len()).size() // VarInt size for script_pubkey
+                    + internal_address_script_pubkey.len()) as u64, // script_pubkey size
+        );
+        let change_cost = calculate_fee(change_weight.to_vbytes_ceil(), long_term_feerate)
+            .map_err(|e| WalletError::General(format!("Fee calculation failed: {}", e)))?;
+        let target_weight = Weight::from_vb_unwrap(
+            (Amount::SIZE
+                + VarInt::from(external_address_script_pubkey.len()).size()
+                + external_address_script_pubkey.len()) as u64,
+        );
+        let avg_output_weight =
+            (change_weight.to_wu() + (target_weight.to_wu() * num_outputs)) / (1 + num_outputs);
+        let avg_input_weight = unspents
+            .iter()
+            .map(|(_, spend_info)| {
+                // Base weight of the input (OutPoint + sequence + empty script_sig)
+                let base_weight = 32 + 4 + 4 + 1; // 32 bytes for OutPoint, 4 bytes for sequence,4 bytes for vout, 1 byte for empty script_sig
+                let witness_weight = spend_info.estimate_witness_size();
+                base_weight + witness_weight as u64
+            })
+            .sum::<u64>()
+            / unspents.len() as u64;
+
+        // Convert UTXOs to OutputGroups
+        let output_groups: Vec<OutputGroup> = unspents
+            .iter()
+            .map(|(utxo, spend_info)| OutputGroup {
+                value: utxo.amount.to_sat(),
+                weight: 36 + spend_info.estimate_witness_size() as u64,
+                input_count: 1,
+                creation_sequence: None,
+            })
+            .collect();
```

Comment:
Add a TODO here for including all utxos of same address in one output group. This reduces the privacy leak for accidental address reuse.

We can later have more complex grouping policy. For now this is fine. Add a TODO for tracking.